Framework/Algorithms/src/ConvertUnits.cpp

```diff
@@ -322,32 +315,93 @@ void ConvertUnits::convertQuickly(API::MatrixWorkspace_sptr outputWS,
 
   if (m_inputEvents)
     eventWS->clearMRU();
+
+  return outputWS;
+}
+
+/** Get the L2, theta and efixed values for a workspace index
+* @param outputUnit :: The output unit
+* @param source :: The source of the instrument
+* @param sample :: The sample of the instrument
+* @param l1 :: The source to sample distance
+* @param emode :: The energy mode
+* @param ws :: The workspace
+* @param thetaFunction :: The function to calculate twotheta
+* @param wsIndex :: The workspace index
+* @param efixed :: the returned fixed energy
+* @param l2 :: The returned sample - detector distance
+* @param twoTheta :: the returned two theta angle
+* @returns true if lookup successful, false on error
+*/
+bool ConvertUnits::getDetectorValues(
+    Kernel::Unit_const_sptr outputUnit, Geometry::IComponent_const_sptr source,
+    Geometry::IComponent_const_sptr sample, double l1, int emode,
+    MatrixWorkspace_const_sptr ws,
+    function<double(const Geometry::IDetector &)> thetaFunction,
+    int64_t wsIndex, double &efixed, double &l2, double &twoTheta) {
```

Comment:
I like that this was extracted into a function. Most of the arguments should not be `shared_ptr` however, since the function is of course not taking ownership. So `source`, `sample`, and `ws` should be passed by `const &` instead.
Since `getDetectorValues` is called once for each spectrum this may actually also make a (very small) performance difference.